server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java

```diff
@@ -87,18 +88,12 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
         }
         if (node.node() != null) {
             assert Strings.hasLength(node.node().getHostAddress()) : node;
-            for (RoutingNode checkNode : allocation.routingNodes()) {
-                if (checkNode.node() == null) {
-                    continue;
-                }
+            for (ShardRouting assignedShard : assignedShards) {
+                DiscoveryNode checkNode = allocation.nodes().get(assignedShard.currentNodeId());
                 // check if its on the same host as the one we want to allocate to
-                assert Strings.hasLength(checkNode.node().getHostAddress()) : checkNode;
-                if (checkNode.node().getHostAddress().equals(node.node().getHostAddress())) {
-                    for (ShardRouting assignedShard : assignedShards) {
-                        if (checkNode.nodeId().equals(assignedShard.currentNodeId())) {
-                            return allocation.debugDecision() ? debugNoAlreadyAllocatedToHost(node, allocation) : Decision.NO;
-                        }
-                    }
+                assert Strings.hasLength(checkNode.getHostAddress()) : checkNode;
```

Comment:
We've lost a `null` check on `checkNode` along the way. I think it can't be `null` any more but I'd like to `assert checkNode != null` just to be sure.